## MIE1624-Course Project-Recommender System (Group4)

*Code ref:* https://www.analyticsvidhya.com/blog/2018/06/comprehensive-guide-recommendation-engine-python/?#

In [37]:
!pip install turicreate

Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.


In [59]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import turicreate
from sklearn.metrics import mean_squared_error
from collections import Counter

In [39]:
df_train = pd.read_csv("train.csv")

In [40]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 11 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   overall         150000 non-null  float64
 1   reviewTime      150000 non-null  object 
 2   reviewerID      150000 non-null  object 
 3   reviewText      149972 non-null  object 
 4   summary         149971 non-null  object 
 5   unixReviewTime  150000 non-null  int64  
 6   category        150000 non-null  object 
 7   price           150000 non-null  object 
 8   itemID          150000 non-null  object 
 9   reviewHash      150000 non-null  int64  
 10  image           351 non-null     object 
dtypes: float64(1), int64(2), object(8)
memory usage: 12.6+ MB


In [41]:
r_cols = ["reviewerID", "itemID", "overall"]
dataset = df_train[r_cols]

# Remove "u" in the column "reviewerID" and "p" in the column "itemID"
dataset["reviewerID"] = dataset["reviewerID"].str.replace("u" ,"")
dataset["reviewerID"] = pd.to_numeric(dataset["reviewerID"])
dataset["itemID"] = dataset["itemID"].str.replace("p" ,"")
dataset["itemID"] = pd.to_numeric(dataset["itemID"])
dataset

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-doc

reviewerID    itemID  overall
0         92990698  23649501      5.0
1         36200649  58458313      5.0
2         10721702  97027626      5.0
3         86003775  43167086      5.0
4         25672859  94494236      5.0
...            ...       ...      ...
149995    29136602  26326919      5.0
149996    51715193  69247882      5.0
149997    28696060   7028930      5.0
149998     4925906  38606785      5.0
149999    79459462   7495887      5.0

[150000 rows x 3 columns]

Split the training set and test (validation) set.

In [42]:
X = dataset[["reviewerID", "itemID"]]
y = dataset["overall"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, 
                                                    random_state=42)

In [43]:
print("In the training set:")
n_reviewers_train = X_train["reviewerID"].unique().shape[0]
print("Number of reviewers: %d." % n_reviewers_train)
n_items_train = X_train["itemID"].unique().shape[0]
print("Number of items: %d." % n_items_train)

In the training set:
Number of reviewers: 51727.
Number of items: 22821.


In [44]:
train_set = X_train
train_set["overall"] = y_train
train_set

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


reviewerID    itemID  overall
27822     36062488  95953136      5.0
85202     80160344   3910736      1.0
42217     10278246  36753582      4.0
119958    65852773  79839929      5.0
37678     61404159  96488536      5.0
...            ...       ...      ...
119879    78315757  43416622      5.0
103694    94793406  71959028      4.0
131932    98922311  24764686      5.0
146867    71705099  73254534      3.0
121958    99996166   2981415      4.0

[105000 rows x 3 columns]

In [45]:
print("In the test set:")
n_reviewers_test = X_test["reviewerID"].unique().shape[0]
print("Number of reviewers: %d." % n_reviewers_test)
n_items_test = X_test["itemID"].unique().shape[0]
print("Number of items: %d." % n_items_test)

In the test set:
Number of reviewers: 29160.
Number of items: 16509.


In [98]:
test_set = X_test
test_set["overall"] = y_test
test_set

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


reviewerID    itemID  overall
59770     27795695  87595611      5.0
21362     25588969  54328640      5.0
127324    32826075  98585649      5.0
140509    54386356  79496784      5.0
144297    54095284  12135090      5.0
...            ...       ...      ...
89890     28436262  43647333      5.0
94082     53116995  11683810      5.0
29674     92859373  62568296      4.0
92625     46206709  41723559      5.0
98351     32997826  56091313      5.0

[45000 rows x 3 columns]

In [47]:
train_data = turicreate.SFrame(train_set)
test_data = turicreate.SFrame(test_set)

# Popularity model

* **train_data**: the dataset to use for training the model
* **user_id**: the column name which represents each reviewer ID
* **item_id**: the column name which represents each item ID to be recommended 
* **target**: the column name representing ratings given by the reviewers

In [48]:
popularity_model = turicreate.popularity_recommender.create(train_data, 
                                                            user_id="reviewerID", 
                                                            item_id="itemID", 
                                                            target="overall")

Preparing data set.

Data has 105000 observations with 51727 users and 22821 items.

Data prepared in: 0.175781s

105000 observations to process; with 22821 unique items.

In [49]:
# Recommend the top 2 items for each reviewer
popularity_recomm = popularity_model.recommend(users=test_data["reviewerID"], k=2)
popularity_recomm.print_rows(num_rows=20)

recommendations finished on 1000/45000 queries. users per second: 4985.91

recommendations finished on 2000/45000 queries. users per second: 5036.51

recommendations finished on 3000/45000 queries. users per second: 5145.47

recommendations finished on 4000/45000 queries. users per second: 5112.25

recommendations finished on 5000/45000 queries. users per second: 5139.31

recommendations finished on 6000/45000 queries. users per second: 5122.09

recommendations finished on 7000/45000 queries. users per second: 5161.94

recommendations finished on 8000/45000 queries. users per second: 5185.22

recommendations finished on 9000/45000 queries. users per second: 5163.88

recommendations finished on 10000/45000 queries. users per second: 5173.83

recommendations finished on 11000/45000 queries. users per second: 5169.34

recommendations finished on 12000/45000 queries. users per second: 5159.97

recommendations finished on 13000/45000 queries. users per second: 5179.49

recommendations finished on 14000/45000 queries. users per second: 5188.56

recommendations finished on 15000/45000 queries. users per second: 5197.95

recommendations finished on 16000/45000 queries. users per second: 5204.22

recommendations finished on 17000/45000 queries. users per second: 5191.58

recommendations finished on 18000/45000 queries. users per second: 5203.25

recommendations finished on 19000/45000 queries. users per second: 5205.5

recommendations finished on 20000/45000 queries. users per second: 5203.77

recommendations finished on 21000/45000 queries. users per second: 5207.43

recommendations finished on 22000/45000 queries. users per second: 5199.97

recommendations finished on 23000/45000 queries. users per second: 5199.67

recommendations finished on 24000/45000 queries. users per second: 5208.32

recommendations finished on 25000/45000 queries. users per second: 5211.54

recommendations finished on 26000/45000 queries. users per second: 5211.38

recommendations finished on 27000/45000 queries. users per second: 5218.63

recommendations finished on 28000/45000 queries. users per second: 5217.25

recommendations finished on 29000/45000 queries. users per second: 5217.63

recommendations finished on 30000/45000 queries. users per second: 5226.13

recommendations finished on 31000/45000 queries. users per second: 5227.56

recommendations finished on 32000/45000 queries. users per second: 5231.48

recommendations finished on 33000/45000 queries. users per second: 5223.27

recommendations finished on 34000/45000 queries. users per second: 5224.96

recommendations finished on 35000/45000 queries. users per second: 5229.51

recommendations finished on 36000/45000 queries. users per second: 5227.51

recommendations finished on 37000/45000 queries. users per second: 5231.95

recommendations finished on 38000/45000 queries. users per second: 5229.93

recommendations finished on 39000/45000 queries. users per second: 5233.42

recommendations finished on 40000/45000 queries. users per second: 5235.21

recommendations finished on 41000/45000 queries. users per second: 5236.8

recommendations finished on 42000/45000 queries. users per second: 5230.23

recommendations finished on 43000/45000 queries. users per second: 5230.43

recommendations finished on 44000/45000 queries. users per second: 5229.58

recommendations finished on 45000/45000 queries. users per second: 5230.21

+------------+----------+-------+------+
| reviewerID |  itemID  | score | rank |
+------------+----------+-------+------+
|  27795695  | 79839929 |  5.0  |  1   |
|  27795695  | 95953136 |  5.0  |  2   |
|  25588969  | 79839929 |  5.0  |  1   |
|  25588969  | 95953136 |  5.0  |  2   |
|  32826075  | 79839929 |  5.0  |  1   |
|  32826075  | 95953136 |  5.0  |  2   |
|  54386356  | 79839929 |  5.0  |  1   |
|  54386356  | 95953136 |  5.0  |  2   |
|  54095284  | 79839929 |  5.0  |  1   |
|  54095284  | 95953136 |  5.0  |  2   |
|  98570622  | 79839929 |  5.0  |  1   |
|  98570622  | 95953136 |  5.0  |  2   |
|  97457161  | 79839929 |  5.0  |  1   |
|  97457161  | 95953136 |  5.0  |  2   |
|  57064374  | 79839929 |  5.0  |  1   |
|  57064374  | 95953136 |  5.0  |  2   |
|  35175691  | 79839929 |  5.0  |  1   |
|  35175691  | 95953136 |  5.0  |  2   |
|  81555857  | 79839929 |  5.0  |  1   |
|  81555857  | 95953136 |  5.0  |  2   |
+------------+----------+-------+------+
[90000 rows x 4 

# Collaborative filtering model

Based on item-item similarities.

Personalization exists in the recommendations.

In [50]:
#Training the model
item_similarity_model = turicreate.item_similarity_recommender.create(train_data, 
                                                                      user_id="reviewerID", 
                                                                      item_id="itemID", 
                                                                      target="overall", 
                                                                      similarity_type="cosine")

# Recommend 2 items for each reviewer
item_similarity_recomm = item_similarity_model.recommend(users=test_data["reviewerID"], k=2)
item_similarity_recomm.print_rows(num_rows=20)

Preparing data set.

Data has 105000 observations with 51727 users and 22821 items.

Data prepared in: 0.175949s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 1.533ms                        | 1.75       |

| 13.01ms                        | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 3.09s                               | 0.5              | 136             |

| 9.68s                               | 100              | 22821           |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 10.7258s

recommendations finished on 1000/45000 queries. users per second: 4064.61

recommendations finished on 2000/45000 queries. users per second: 3979.24

recommendations finished on 3000/45000 queries. users per second: 3969.26

recommendations finished on 4000/45000 queries. users per second: 3964.23

recommendations finished on 5000/45000 queries. users per second: 3965.7

recommendations finished on 6000/45000 queries. users per second: 3989.13

recommendations finished on 7000/45000 queries. users per second: 3999.74

recommendations finished on 8000/45000 queries. users per second: 3990.32

recommendations finished on 9000/45000 queries. users per second: 3983.41

recommendations finished on 10000/45000 queries. users per second: 3980.21

recommendations finished on 11000/45000 queries. users per second: 3981.93

recommendations finished on 12000/45000 queries. users per second: 3978.77

recommendations finished on 13000/45000 queries. users per second: 3951.11

recommendations finished on 14000/45000 queries. users per second: 3946.63

recommendations finished on 15000/45000 queries. users per second: 3947.78

recommendations finished on 16000/45000 queries. users per second: 3939.84

recommendations finished on 17000/45000 queries. users per second: 3944.43

recommendations finished on 18000/45000 queries. users per second: 3950.26

recommendations finished on 19000/45000 queries. users per second: 3960.54

recommendations finished on 20000/45000 queries. users per second: 3960.23

recommendations finished on 21000/45000 queries. users per second: 3964.01

recommendations finished on 22000/45000 queries. users per second: 3967.92

recommendations finished on 23000/45000 queries. users per second: 3972.22

recommendations finished on 24000/45000 queries. users per second: 3970.15

recommendations finished on 25000/45000 queries. users per second: 3972.19

recommendations finished on 26000/45000 queries. users per second: 3974.82

recommendations finished on 27000/45000 queries. users per second: 3979.79

recommendations finished on 28000/45000 queries. users per second: 3981.5

recommendations finished on 29000/45000 queries. users per second: 3983.7

recommendations finished on 30000/45000 queries. users per second: 3988.05

recommendations finished on 31000/45000 queries. users per second: 3991.12

recommendations finished on 32000/45000 queries. users per second: 3987.83

recommendations finished on 33000/45000 queries. users per second: 3986.32

recommendations finished on 34000/45000 queries. users per second: 3989.44

recommendations finished on 35000/45000 queries. users per second: 3992.12

recommendations finished on 36000/45000 queries. users per second: 3994.26

recommendations finished on 37000/45000 queries. users per second: 3993.67

recommendations finished on 38000/45000 queries. users per second: 3995.85

recommendations finished on 39000/45000 queries. users per second: 3998.3

recommendations finished on 40000/45000 queries. users per second: 3998.21

recommendations finished on 41000/45000 queries. users per second: 3996.95

recommendations finished on 42000/45000 queries. users per second: 3998.38

recommendations finished on 43000/45000 queries. users per second: 3998.98

recommendations finished on 44000/45000 queries. users per second: 4001.45

recommendations finished on 45000/45000 queries. users per second: 3998.59

+------------+----------+----------------------+------+
| reviewerID |  itemID  |        score         | rank |
+------------+----------+----------------------+------+
|  27795695  | 22205528 |  0.5277504026889801  |  1   |
|  27795695  | 85170604 | 0.25509774684906006  |  2   |
|  25588969  | 82957180 |  0.8627462387084961  |  1   |
|  25588969  | 95155072 |       0.78125        |  2   |
|  32826075  | 63144896 | 0.010032687187194824 |  1   |
|  32826075  | 5640389  | 0.009872047901153565 |  2   |
|  54386356  | 63144896 | 0.010032687187194824 |  1   |
|  54386356  | 5640389  | 0.009872047901153565 |  2   |
|  54095284  | 74936515 | 0.44305071234703064  |  1   |
|  54095284  | 24039227 | 0.41136234998703003  |  2   |
|  98570622  | 86142828 |        0.625         |  1   |
|  98570622  | 26643528 |        0.625         |  2   |
|  97457161  | 92296104 | 0.13122057035321094  |  1   |
|  97457161  | 91265196 | 0.09973868287977625  |  2   |
|  57064374  | 63144896 | 0.010032687187194824 |

# Max factorization

In [80]:
class max_factorization():

    # Initialization
    def __init__(self, R, K, alpha, beta, iterations):
        self.R = R                           # reviewer-item rating matrix
        self.num_reviewers, self.num_items = R.shape
        self.K = K                           # number of latent features
        self.alpha = alpha                   # learning rate
        self.beta = beta                     # regularization param
        self.iterations = iterations         # number of iterations

    def train(self):
        # self.P: the user(reviewer)-feature affinity matrix
        # self.Q: the item-feature relevance matrix
        self.P = np.random.normal(scale=1./self.K, size=(self.num_reviewers, self.K))
        self.Q = np.random.normal(scale=1./self.K, size=(self.num_items, self.K))

        # Initializing the bias terms
        self.b_reviewer = np.zeros(self.num_reviewers)
        self.b_item = np.zeros(self.num_items)
        self.b = np.mean(self.R[np.where(self.R != 0)])

        # List of training samples
        self.samples = []
        for i in range(self.num_reviewers):
            for j in range(self.num_items):
                if self.R[i, j] != 0:
                    self.samples.append((i, j, self.R[i, j]))
                    
        # Perform stochastic gradient descent 
        training_process = []
        for iter in range(self.iterations):
            if (iter + 1) % 20 == 0:
                print("Iteration: ", (iter + 1))
            np.random.shuffle(self.samples)
            self.sgd()
            MSE = self.compute_mse()
            training_process.append((iter, MSE))
            if (iter + 1) % 10 == 0:
                print("Training MSE = %.3f." % MSE)
                print()
        return training_process

    # Computing total mean squared error
    def compute_mse(self):
        xs, ys = self.R.nonzero()  # the locations of non-zero values in R
        predicted = self.full_matrix()
        squared_errors = []
        for x, y in zip(xs, ys):
            squared_errors.append((self.R[x, y] - predicted[x, y])**2)
        return np.mean(squared_errors)

    # Stochastic gradient descent to find the optimized P and Q 
    def sgd(self):
        for i, j, r in self.samples:  # recall what we have appended in self.samples
            prediction = self.predict_rating(i, j)
            e = r - prediction # the error term

            self.b_reviewer[i] += self.alpha * (e - self.beta * self.b_reviewer[i])
            self.b_item[j] += self.alpha * (e - self.beta * self.b_item[j])

            self.P[i, :] += self.alpha * (e * self.Q[j, :] - self.beta * self.P[i,:])
            self.Q[j, :] += self.alpha * (e * self.P[i, :] - self.beta * self.Q[j,:])

    # Get rating for a reviewer i and item j pair
    def predict_rating(self, i, j):
        prediction = self.b + self.b_reviewer[i] + self.b_item[j] + self.P[i, :].dot(self.Q[j, :].T)
        return prediction

    # Get the full reviewer-item rating matrix
    def full_matrix(self):
        full_pred = self.b + self.b_reviewer[:, np.newaxis] + self.b_item[np.newaxis:,] + self.P.dot(self.Q.T)
        return full_pred

In [99]:
test_set_to_pred = test_set.copy()
test_set_to_pred["overall"] = 0
test_set_to_pred

reviewerID    itemID  overall
59770     27795695  87595611        0
21362     25588969  54328640        0
127324    32826075  98585649        0
140509    54386356  79496784        0
144297    54095284  12135090        0
...            ...       ...      ...
89890     28436262  43647333        0
94082     53116995  11683810        0
29674     92859373  62568296        0
92625     46206709  41723559        0
98351     32997826  56091313        0

[45000 rows x 3 columns]

In [100]:
# Create a subset of train_set and test_set for illustration
subset = train_set[:7000].append(test_set_to_pred[:3000])
subset

reviewerID    itemID  overall
27822     36062488  95953136      5.0
85202     80160344   3910736      1.0
42217     10278246  36753582      4.0
119958    65852773  79839929      5.0
37678     61404159  96488536      5.0
...            ...       ...      ...
146364    82407503  73234226      0.0
70240     67260809  70858769      0.0
114852    82465038  49079644      0.0
132680    91558099  63232619      0.0
116391    36756458  84130808      0.0

[10000 rows x 3 columns]

In [101]:
print("In the subset:")
n_reviewers_sub = subset["reviewerID"].unique().shape[0]
print("Number of reviewers: %d." % n_reviewers_sub)
n_items_sub = subset["itemID"].unique().shape[0]
print("Number of items: %d." % n_items_sub)

In the subset:
Number of reviewers: 8438.
Number of items: 6705.


Convert the reviewer-item rating to a matrix, i.e., get the R matrix.

pivot() returns reshaped DataFrame organized by given index / column values.

In [93]:
R_df = subset.pivot(index="reviewerID", columns="itemID", values="overall")
R = np.array(R_df)
R[np.isnan(R)]=0  # replace all nans with 0
print("Shape of R:", R.shape)
print("Matrix R:")
print(R)

Shape of R: (8438, 6705)
Matrix R:
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


In [102]:
# Count the values in matrix R
key = np.unique(R)
result = {}
for k in key:
    mask = (R == k)
    R_new = R[mask]
    v = R_new.size
    result[k] = v
print(result)

{0.0: 56569790, 1.0: 211, 2.0: 247, 3.0: 572, 4.0: 1318, 5.0: 4652}


We have in total 8438*6705 reviewer-item pairs (ratings). But from the results of counting, we can see that the majority of ratings are missing.

Start filling in the full matrix

In [91]:
mf = max_factorization(R, K=20, alpha=0.001, beta=0.01, iterations=500)
training_process = mf.train()
print("P x Q:")
print(mf.full_matrix())
print("Rounded P x Q:")
print(np.round(mf.full_matrix()))
print()

Training MSE = 0.935.

Iteration:  20
Training MSE = 0.892.

Training MSE = 0.852.

Iteration:  40
Training MSE = 0.815.

Training MSE = 0.780.

Iteration:  60
Training MSE = 0.747.

Training MSE = 0.715.

Iteration:  80
Training MSE = 0.686.

Training MSE = 0.658.

Iteration:  100
Training MSE = 0.632.

Training MSE = 0.606.

Iteration:  120
Training MSE = 0.583.

Training MSE = 0.560.

Iteration:  140
Training MSE = 0.538.

Training MSE = 0.518.

Iteration:  160
Training MSE = 0.498.

Training MSE = 0.479.

Iteration:  180
Training MSE = 0.461.

Training MSE = 0.444.

Iteration:  200
Training MSE = 0.428.

Training MSE = 0.412.

Iteration:  220
Training MSE = 0.398.

Training MSE = 0.383.

Iteration:  240
Training MSE = 0.369.

Training MSE = 0.356.

Iteration:  260
Training MSE = 0.344.

Training MSE = 0.332.

Iteration:  280
Training MSE = 0.320.

Training MSE = 0.309.

Iteration:  300
Training MSE = 0.298.

Training MSE = 0.288.

Iteration:  320
Training MSE = 0.278.

Training MSE

Compute the test MSE

In [96]:
rounded_predictions = pd.DataFrame(np.round(mf.full_matrix()))
rounded_predictions.columns = R_df.columns
rounded_predictions.index = R_df.index
rounded_predictions

itemID      1074      40836     58069     ...  99938488  99952270  99987089
reviewerID                                ...                              
4783             5.0       4.0       5.0  ...       5.0       5.0       4.0
9410             5.0       4.0       5.0  ...       5.0       4.0       4.0
21805            4.0       4.0       4.0  ...       4.0       4.0       4.0
34344            4.0       4.0       4.0  ...       4.0       4.0       4.0
36004            5.0       4.0       5.0  ...       5.0       4.0       4.0
...              ...       ...       ...  ...       ...       ...       ...
99979017         4.0       4.0       4.0  ...       4.0       4.0       4.0
99983493         5.0       4.0       5.0  ...       5.0       4.0       4.0
99986481         5.0       4.0       5.0  ...       5.0       4.0       4.0
99992203         4.0       4.0       4.0  ...       4.0       4.0       4.0
99996166         4.0       3.0       4.0  ...       4.0       3.0       3.0

[8438 rows x 6705 columns]

In [103]:
test_set[:3000]

reviewerID    itemID  overall
59770     27795695  87595611      5.0
21362     25588969  54328640      5.0
127324    32826075  98585649      5.0
140509    54386356  79496784      5.0
144297    54095284  12135090      5.0
...            ...       ...      ...
146364    82407503  73234226      5.0
70240     67260809  70858769      5.0
114852    82465038  49079644      5.0
132680    91558099  63232619      5.0
116391    36756458  84130808      4.0

[3000 rows x 3 columns]

In [104]:
test_MSE_list = []
for row in range(3000):
    test_pair = test_set.iloc[row]
    reviewer = test_pair["reviewerID"]
    item = test_pair["itemID"]
    true_rating = test_pair["overall"]
    pred_rating = rounded_predictions.loc[reviewer, item]
    pair_squared_error = (true_rating - pred_rating)**2
    test_MSE_list.append(pair_squared_error)
print("The test MSE is %0.3f." % np.mean(test_MSE_list))

The test MSE is 1.189.
